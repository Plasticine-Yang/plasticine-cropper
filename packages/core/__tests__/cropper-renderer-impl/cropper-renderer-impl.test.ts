import {
  type CropperRenderer,
  CropperRendererImpl,
  CROPPER_LINE_NOT_RESIZABLE,
  CROPPER_POINT_NOT_RESIZABLE,
} from '@internal-dist/index'

describe('cropper renderer impl', () => {
  const setup = () => {
    const containerElement = document.createElement('div')
    const imageElement = document.createElement('img')

    containerElement.appendChild(imageElement)

    const cropperRenderer: CropperRenderer = new CropperRendererImpl(imageElement)

    return {
      cropperRenderer,
      imageElement,
    }
  }

  test('should throw error without container', () => {
    const imageElement = document.createElement('img')

    const internalTest = () => {
      new CropperRendererImpl(imageElement)
    }
    expect(internalTest).toThrowErrorMatchingInlineSnapshot('"未发现图片的容器元素，无法渲染 plasticine cropper"')
  })

  test('should get all cropper elements', () => {
    const { cropperRenderer } = setup()

    expect(cropperRenderer.getCropperElements()).toMatchSnapshot()
  })

  test('should toggle visibility of raw image element', () => {
    const { cropperRenderer, imageElement } = setup()

    // hide
    cropperRenderer.hideRawImageElement()
    expect(Array.from(imageElement.classList)).toMatchInlineSnapshot(`
      [
        "plasticine-cropper-common-hidden",
      ]
    `)

    // show
    cropperRenderer.showRawImageElement()
    expect(Array.from(imageElement.classList)).toMatchInlineSnapshot('[]')
  })

  test('should move crop container', () => {
    const { cropperRenderer } = setup()
    const { cropContainer } = cropperRenderer.getCropperElements()

    cropperRenderer.moveCropContainer(400, 300)

    expect({ top: cropContainer.style.top, left: cropContainer.style.left }).toMatchInlineSnapshot(`
      {
        "left": "400px",
        "top": "300px",
      }
    `)
  })

  test('should control resizable of crop container', () => {
    const { cropperRenderer } = setup()
    const { cropContainerLines, cropContainerPoints } = cropperRenderer.getCropperElements()

    cropperRenderer.makeCropContainerResizable()

    for (const lineElement of Object.values(cropContainerLines)) {
      expect(lineElement.classList.contains(CROPPER_LINE_NOT_RESIZABLE)).toBe(false)
    }

    for (const pointElement of Object.values(cropContainerPoints)) {
      expect(pointElement.classList.contains(CROPPER_POINT_NOT_RESIZABLE)).toBe(false)
    }

    cropperRenderer.makeCropContainerNotResizable()

    for (const lineElement of Object.values(cropContainerLines)) {
      expect(lineElement.classList.contains(CROPPER_LINE_NOT_RESIZABLE)).toBe(true)
    }

    for (const pointElement of Object.values(cropContainerPoints)) {
      expect(pointElement.classList.contains(CROPPER_POINT_NOT_RESIZABLE)).toBe(true)
    }
  })
})
