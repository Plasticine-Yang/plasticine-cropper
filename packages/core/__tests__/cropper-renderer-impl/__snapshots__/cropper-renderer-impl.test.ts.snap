// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cropper renderer impl > should get all cropper elements 1`] = `
{
  "cropContainer": <div
    class="plasticine-cropper__crop-container"
  >
    <div
      class="plasticine-cropper__view-container"
    />
    <div
      class="plasticine-cropper__dotted-line plasticine-cropper__dotted-line--horizontal"
    />
    <div
      class="plasticine-cropper__dotted-line plasticine-cropper__dotted-line--vertical"
    />
    <div
      class="plasticine-cropper__focus"
    />
    <div
      class="plasticine-cropper__line plasticine-cropper__line--n"
    />
    <div
      class="plasticine-cropper__line plasticine-cropper__line--e"
    />
    <div
      class="plasticine-cropper__line plasticine-cropper__line--s"
    />
    <div
      class="plasticine-cropper__line plasticine-cropper__line--w"
    />
    <div
      class="plasticine-cropper__point plasticine-cropper__point--n"
    />
    <div
      class="plasticine-cropper__point plasticine-cropper__point--e"
    />
    <div
      class="plasticine-cropper__point plasticine-cropper__point--s"
    />
    <div
      class="plasticine-cropper__point plasticine-cropper__point--w"
    />
    <div
      class="plasticine-cropper__point plasticine-cropper__point--ne"
    />
    <div
      class="plasticine-cropper__point plasticine-cropper__point--nw"
    />
    <div
      class="plasticine-cropper__point plasticine-cropper__point--sw"
    />
    <div
      class="plasticine-cropper__point plasticine-cropper__point--se"
    />
  </div>,
  "cropContainerLines": {
    "e": <div
      class="plasticine-cropper__line plasticine-cropper__line--e"
    />,
    "n": <div
      class="plasticine-cropper__line plasticine-cropper__line--n"
    />,
    "s": <div
      class="plasticine-cropper__line plasticine-cropper__line--s"
    />,
    "w": <div
      class="plasticine-cropper__line plasticine-cropper__line--w"
    />,
  },
  "cropContainerPoints": {
    "e": <div
      class="plasticine-cropper__point plasticine-cropper__point--e"
    />,
    "n": <div
      class="plasticine-cropper__point plasticine-cropper__point--n"
    />,
    "ne": <div
      class="plasticine-cropper__point plasticine-cropper__point--ne"
    />,
    "nw": <div
      class="plasticine-cropper__point plasticine-cropper__point--nw"
    />,
    "s": <div
      class="plasticine-cropper__point plasticine-cropper__point--s"
    />,
    "se": <div
      class="plasticine-cropper__point plasticine-cropper__point--se"
    />,
    "sw": <div
      class="plasticine-cropper__point plasticine-cropper__point--sw"
    />,
    "w": <div
      class="plasticine-cropper__point plasticine-cropper__point--w"
    />,
  },
  "root": <div
    class="plasticine-cropper"
  >
    <div
      class="plasticine-cropper__overlay"
    />
    <div
      class="plasticine-cropper__canvas"
    />
    <div
      class="plasticine-cropper__crop-container"
    >
      <div
        class="plasticine-cropper__view-container"
      />
      <div
        class="plasticine-cropper__dotted-line plasticine-cropper__dotted-line--horizontal"
      />
      <div
        class="plasticine-cropper__dotted-line plasticine-cropper__dotted-line--vertical"
      />
      <div
        class="plasticine-cropper__focus"
      />
      <div
        class="plasticine-cropper__line plasticine-cropper__line--n"
      />
      <div
        class="plasticine-cropper__line plasticine-cropper__line--e"
      />
      <div
        class="plasticine-cropper__line plasticine-cropper__line--s"
      />
      <div
        class="plasticine-cropper__line plasticine-cropper__line--w"
      />
      <div
        class="plasticine-cropper__point plasticine-cropper__point--n"
      />
      <div
        class="plasticine-cropper__point plasticine-cropper__point--e"
      />
      <div
        class="plasticine-cropper__point plasticine-cropper__point--s"
      />
      <div
        class="plasticine-cropper__point plasticine-cropper__point--w"
      />
      <div
        class="plasticine-cropper__point plasticine-cropper__point--ne"
      />
      <div
        class="plasticine-cropper__point plasticine-cropper__point--nw"
      />
      <div
        class="plasticine-cropper__point plasticine-cropper__point--sw"
      />
      <div
        class="plasticine-cropper__point plasticine-cropper__point--se"
      />
    </div>
  </div>,
}
`;
